src/data/list/pi.lean

```diff
@@ -0,0 +1,45 @@
+/-
+Copyright (c) 2023 Yuyang Zhao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yuyang Zhao
+-/
+import tactic.ext
+import tactic.split_ifs
+
+/-!
+# The cartesian product of lists
+-/
+
+namespace list
+variables {ι : Type*} [decidable_eq ι] {α : ι → Sort*}
+
+namespace pi
+
+/-- Given `α : ι → Sort*`, `pi.nil α` is the trivial dependent function out of the empty list. -/
+def nil {ι : Type*} (α : ι → Sort*) : (Π i ∈ ([] : list ι), α i).
+
+variables {i : ι} {l : list ι}
+
+/-- Given `α : ι → Sort*`, a list `l` and a term `i`, as well as a term `a : α i` and a
+function `f` such that `f j : α j` for all `i'` in `m`, `pi.cons a f` is a function `g` such
+that `g i'' : α i''` for all `i''` in `i :: l`. -/
+def cons (a : α i) (f : Π j ∈ l, α j) : Π j ∈ (i :: l), α j :=
```

Comment:
`l` should be explicit to match the multiset version

Reply:
In most cases it can be inferred automatically. Is there any disadvantage to making it implicit in `multiset.pi.cons`?

Reply:
Yes, it avoid making large changes to `multiset.pi.cons` which is already a ported file.